packages/lit-html/src/directives/async-replace.ts

```diff
@@ -10,69 +10,127 @@ import {AsyncDirective} from '../async-directive.js';
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
   render<T>(value: AsyncIterable<T>, _mapper?: Mapper<T>) {
     return noChange;
   }
 
-  update(_part: ChildPart, [value, mapper]: DirectiveParameters<this>) {
+  update(_part: ChildPart, [iterable, mapper]: DirectiveParameters<this>) {
     // If we've already set up this particular iterable, we don't need
     // to do anything.
-    if (value === this._value) {
+    if (iterable === this.__iterable) {
       return;
     }
-    this._value = value;
-    this.__iterate(mapper);
+    // If we're starting a new promise, stop tracking the last one, since its
+    // result is no longer relevant
+    if (this.__nextPromise !== undefined) {
+      promiseToDirectiveMap.delete(this.__nextPromise);
+    }
+    this.__mapper = mapper;
+    this.__iterable = iterable;
+    this.index = 0;
+    // Note, we use the iterator protocol manually (rather than for/await/of) so
+    // that we can break the reference between the next promise and the directive
+    // when the directive disconnects, and be able to resume iteration if the
+    // directive reconnects. The only way to stop iteration and allow the
+    // directive (and all its DOM) to be gc'ed via for/await/of would be to break
+    // out of the loop, which precludes resuming the iteration if/when the
+    // directive is reconnected.
+    this.__iterator = iterable[Symbol.asyncIterator]();
+    this.__awaitNextValue();
     return noChange;
   }
 
-  // Separate function to avoid an iffe in update; update can't be async
-  // because its return value must be `noChange`
-  private async __iterate(mapper?: Mapper<unknown>) {
-    let i = 0;
-    const {_value: value} = this;
-    for await (let v of value!) {
-      // Check to make sure that value is the still the current value of
-      // the part, and if not bail because a new value owns this part
-      if (this._value !== value) {
-        break;
-      }
-
-      // If we were disconnected, pause until reconnected
-      if (this._reconnectPromise) {
-        await this._reconnectPromise;
-      }
-
-      // As a convenience, because functional-programming-style
-      // transforms of iterables and async iterables requires a library,
-      // we accept a mapper function. This is especially convenient for
-      // rendering a template for each item.
-      if (mapper !== undefined) {
-        v = mapper(v, i);
+  private __awaitNextValue(nextPromise = this.__iterator!.next()) {
+    this.__nextPromise = nextPromise;
+    if (this.isConnected) {
+      // We still await the nextPromise even when disconnected (since if the
+      // directive reconnects it will need to handle the result), but only
+      // associate the directive to the promise when connected
+      promiseToDirectiveMap.set(nextPromise, this);
+    }
+    nextPromise.then((result) => {
+      const directive = promiseToDirectiveMap.get(nextPromise);
```

Comment:
Here `directive` is always the same as `this` isn't it? Does this mean we don't need to actually store the directive value and only need to know if this promise is mapped?

Reply:
The whole thing is not closing over `this` in the thenFn. Going through the WeakMap with the promise as the key is what breaks the dependency.